ContextClosedEvent on child context shuts down grails

### Task List

- [ X] Steps to reproduce provided
- [X ] Stacktrace (if present) provided
- [X ] Example that reproduces the problem uploaded to Github
- [ X] Full description of the issue provided (see below)

### Steps to Reproduce

1. Create and load a child application context in a service method (or anywhere in a Grails App):
```
ClassPathResource myAppCtx = new ClassPathResource("AppCtx.xml")
ApplicationContext context = new GenericXmlApplicationContext()
context.setParent(Holders.grailsApplication.mainContext)
context.load(myAppCtx)
context.refresh()
```
2. Close the child context
`context.close()`

3. Access the grailsApplication bean via Holders and watch it blow up due to "java.lang.IllegalArgumentException: GrailsApplication not found"

`log.info("Grails App still alive: ${Holders.grailsApplication != null}")`

Exception:
```
Caused by: java.lang.IllegalArgumentException: GrailsApplication not found
	at org.springframework.util.Assert.notNull(Assert.java:134)
	at grails.util.Holders.getGrailsApplication(Holders.java:129)

```
### Expected Behaviour

Applications should be able to load their own application contexts that are children of the main Grails application context. They should be able to manage the lifecycle of these contexts without impacting the main application. This includes the ability to close their child application context when they have finished with it without shutting down the application.



### Actual Behaviour

The GrailsApplicationPostProcessor.onApplicationEvent() method doesn't check whether the ContextClosedEvent is for the Grails main context or some other user defined Application Context before shutting down the application.

This means that if applications add their own child Application Contexts and later close them, then the Grails application as a whole is shutdown. the behaviour should be more consistent with Spring where the lifecycle of a child context doesn't impact the parent.

### Environment Information

- **Operating System**: Windows 10
- **Grails Version:** 3.3.6
- **JDK Version:** 1.8.0_181
- **Container Version (If Applicable):** N/A

### Example Application

https://github.com/jchharris/GrailsApplicationPostProcessorBug

--- a/grails-core/src/main/groovy/grails/boot/config/GrailsApplicationPostProcessor.groovy
+++ b/grails-core/src/main/groovy/grails/boot/config/GrailsApplicationPostProcessor.groovy
@@ -235,50 +235,53 @@ class GrailsApplicationPostProcessor implements BeanDefinitionRegistryPostProces
     void onApplicationEvent(ApplicationContextEvent event) {
         ApplicationContext context = event.applicationContext
 
-        Collection<GrailsApplicationLifeCycle> lifeCycleBeans = context.getBeansOfType(GrailsApplicationLifeCycle).values()
-        if(event instanceof ContextRefreshedEvent) {
-            if(context.containsBean("grailsDomainClassMappingContext")) {
-                grailsApplication.setMappingContext(
-                    context.getBean("grailsDomainClassMappingContext", MappingContext)
-                )
-            }
-            Environment.setInitializing(false)
-            pluginManager.setApplicationContext(context)
-            pluginManager.doDynamicMethods()
-            for(GrailsApplicationLifeCycle lifeCycle in lifeCycleBeans) {
-                lifeCycle.doWithDynamicMethods()
-            }
-            pluginManager.doPostProcessing(context)
-            for(GrailsApplicationLifeCycle lifeCycle in lifeCycleBeans) {
-                lifeCycle.doWithApplicationContext()
-            }
-            Holders.pluginManager = pluginManager
-            Map<String,Object> eventMap = [:]
-            eventMap.put('source', pluginManager)
+        if (!applicationContext || applicationContext == context) {
+            // Only act if the event is for our context
+            Collection<GrailsApplicationLifeCycle> lifeCycleBeans = context.getBeansOfType(GrailsApplicationLifeCycle).values()
+            if (event instanceof ContextRefreshedEvent) {
+                if (context.containsBean("grailsDomainClassMappingContext")) {
+                    grailsApplication.setMappingContext(
+                        context.getBean("grailsDomainClassMappingContext", MappingContext)
+                    )
+                }
+                Environment.setInitializing(false)
+                pluginManager.setApplicationContext(context)
+                pluginManager.doDynamicMethods()
+                for (GrailsApplicationLifeCycle lifeCycle in lifeCycleBeans) {
+                    lifeCycle.doWithDynamicMethods()
+                }
+                pluginManager.doPostProcessing(context)
+                for (GrailsApplicationLifeCycle lifeCycle in lifeCycleBeans) {
+                    lifeCycle.doWithApplicationContext()
+                }
+                Holders.pluginManager = pluginManager
+                Map<String, Object> eventMap = [:]
+                eventMap.put('source', pluginManager)
 
-            pluginManager.onStartup(eventMap)
-            for(GrailsApplicationLifeCycle lifeCycle in lifeCycleBeans) {
-                lifeCycle.onStartup(eventMap)
-            }
-        }
-        else if(event instanceof ContextClosedEvent) {
-            pluginManager.shutdown()
-            Map<String,Object> eventMap = [:]
-            eventMap.put('source', pluginManager)
-            for(GrailsApplicationLifeCycle lifeCycle in lifeCycleBeans) {
-                lifeCycle.onShutdown(eventMap)
-            }
-            ShutdownOperations.runOperations()
-            Holders.clear()
-            if(reloadingEnabled) {
-                try {
-                    GrailsSpringLoadedPlugin.unregister()
-                } catch (Throwable e) {
-                    // ignore
+                pluginManager.onStartup(eventMap)
+                for (GrailsApplicationLifeCycle lifeCycle in lifeCycleBeans) {
+                    lifeCycle.onStartup(eventMap)
                 }
             }
+            else if(event instanceof ContextClosedEvent) {
+                pluginManager.shutdown()
+                Map<String, Object> eventMap = [:]
+                eventMap.put('source', pluginManager)
+                for (GrailsApplicationLifeCycle lifeCycle in lifeCycleBeans) {
+                    lifeCycle.onShutdown(eventMap)
+                }
+                ShutdownOperations.runOperations()
+                Holders.clear()
+                if (reloadingEnabled) {
+                    try {
+                        GrailsSpringLoadedPlugin.unregister()
+                    } catch (Throwable e) {
+                        // ignore
+                    }
+                }
 
-            GrailsApp.setDevelopmentModeActive(false)
+                GrailsApp.setDevelopmentModeActive(false)
+            }
         }
     }
 

